packages/@sanity/cli/src/actions/login/login.js

```diff
@@ -23,47 +38,57 @@ export default async function login(args, context) {
   // Get the desired authentication provider
   const provider = await getProvider({client, sso, experimental, output, prompt, specifiedProvider})
   if (provider === undefined) {
-    output.print(chalk.red('No authentication providers found'))
-    return
+    throw new Error('No authentication providers found')
   }
 
-  // Open an authentication listener channel and wait for secret
-  const iv = crypto.randomBytes(8).toString('hex')
-  const es = getAuthChannel(client.config().apiHost, provider, iv)
-  const encryptedToken = getAuthToken(es) // This is a promise, will resolve later
-  const {secret, url} = await getAuthInfo(es)
+  // Initiate local listen server for OAuth callback
+  const apiHost = client.config().apiHost || 'https://api.sanity.io'
+  const {server, token: tokenPromise} = await startServerForTokenCallback({apiHost, apiClient})
 
-  // Open a browser on the login page (or tell the user to)
-  const providerUrl = urlParser.parse(url, true)
-  providerUrl.query.label = `${os.hostname()} / ${os.platform()}`
-  const loginUrl = urlParser.format(providerUrl)
+  const serverUrl = server.address()
+  if (!serverUrl || typeof serverUrl === 'string') {
+    // Note: `serverUrl` is string only when binding to unix sockets.
+    // thus we can safely assume Something is Wrong™ if it's a string
+    throw new Error('Failed to start auth callback server')
+  }
+
+  // Build a login URL that redirects back to OAuth flow on success
+  const loginUrl = new URL(provider.url)
+  const platformName = os.platform()
+  const platform = platformName in platformNames ? platformNames[platformName] : platformName
+  const hostname = os.hostname().replace(/\.(local|lan)$/g, '')
 
+  loginUrl.searchParams.set('type', 'token')
+  loginUrl.searchParams.set('label', `${hostname} / ${platform}`)
+  loginUrl.searchParams.set('origin', `http://localhost:${serverUrl.port}${callbackEndpoint}`)
+
+  // Open a browser on the login page (or tell the user to)
   const shouldLaunchBrowser = canLaunchBrowser() && openFlag !== false
   const actionText = shouldLaunchBrowser ? 'Opening browser at' : 'Please open a browser at'
 
-  output.print(`\n${actionText} ${loginUrl}\n`)
+  output.print(`\n${actionText} ${loginUrl.href}\n`)
   const spin = output
     .spinner('Waiting for browser login to complete... Press Ctrl + C to cancel')
     .start()
 
   if (shouldLaunchBrowser) {
-    open(loginUrl, {wait: false})
+    open(loginUrl.href)
```

Comment:
This still needs the `wait: false` flag, since it is using an old version of the `opn` dependency, which waits by default. Currently the login command hangs after logging in, because it is waiting for the browser to close.

Reply:
ah that makes sense, thanks for catching that!